pkg/service/credential/service.go

```diff
@@ -675,3 +658,71 @@ func (s Service) BatchCreateCredentials(ctx context.Context, batchRequest BatchC
 
 	return credResponse, nil
 }
+
+func (s Service) BatchUpdateCredentialStatus(ctx context.Context, batchRequest BatchUpdateCredentialStatusRequest) (*BatchUpdateCredentialStatusResponse, error) {
+	watchKeys := make([]storage.WatchKey, 0, len(batchRequest.Requests))
+	updateFuncs := make([]storage.BusinessLogicFunc, 0, len(batchRequest.Requests))
+	for _, request := range batchRequest.Requests {
+		statusListCredentialWatchKey, err := s.statusListCredentialWatchKey(ctx, request.ID)
+		if err != nil {
+			return nil, err
+		}
+		watchKeys = append(watchKeys, *statusListCredentialWatchKey)
+
+		slcMetadata := StatusListCredentialMetadata{statusListCredentialWatchKey: *statusListCredentialWatchKey}
+		returnFunc := s.updateCredentialStatusFunc(request, slcMetadata)
+		updateFuncs = append(updateFuncs, returnFunc)
+	}
+	returnValue, err := s.storage.db.Execute(ctx, func(ctx context.Context, tx storage.Tx) (any, error) {
+		batchResponse := BatchUpdateCredentialStatusResponse{
+			CredentialStatuses: make([]Status, 0, len(batchRequest.Requests)),
+		}
+		for i, updateFunc := range updateFuncs {
+			updateResp, err := updateFunc(ctx, tx)
+			if err != nil {
+				return nil, err
+			}
+			batchResponse.CredentialStatuses = append(batchResponse.CredentialStatuses, updateResp.(*UpdateCredentialStatusResponse).Status)
+			batchResponse.CredentialStatuses[i].ID = batchRequest.Requests[i].ID
+		}
+		return &batchResponse, nil
+	}, watchKeys)
+	if err != nil {
+		return nil, errors.Wrap(err, "execute")
+	}
+
+	batchResponse, ok := returnValue.(*BatchUpdateCredentialStatusResponse)
+	if !ok {
+		return nil, errors.New("casting to BatchUpdateCredentialStatusResponse")
+	}
+
+	return batchResponse, nil
+}
+
+func (s Service) statusListCredentialWatchKey(ctx context.Context, id string) (*storage.WatchKey, error) {
+	gotCred, err := s.storage.GetCredential(ctx, id)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading credential")
+	}
+
+	if gotCred.Credential.CredentialStatus == nil {
```

Comment:
I would recommend a new helper like `HasCredentialStatus` which can also check that `gotCred` is not nil and `gotCred.Credential` is not nil, or else we're at risk for NPEs